src/coreclr/src/jit/emit.cpp

```diff
@@ -5320,50 +5328,40 @@ UNATIVE_OFFSET emitter::emitFindOffset(insGroup* ig, unsigned insNum)
  *  block.
  */
 
-UNATIVE_OFFSET emitter::emitDataGenBeg(UNATIVE_OFFSET size, bool align)
+UNATIVE_OFFSET emitter::emitDataGenBeg(UNATIVE_OFFSET size, UNATIVE_OFFSET alignment)
 {
     unsigned     secOffs;
     dataSection* secDesc;
 
     assert(emitDataSecCur == nullptr);
 
-    /* The size better not be some kind of an odd thing */
+    // The size must not be zero and must be a multiple of 4
+    assert((size != 0) && ((size % 4) == 0));
 
-    assert(size && size % sizeof(int) == 0);
+    // This restricts the alignment to: 1, 2, 4, 8, 16, or 32
+    // Alignments greater than 32 would require VM support in ICorJitInfo::allocMem
 
-    /* Get hold of the current offset */
+    const size_t MaxAlignment = 32;
+    assert(isPow2(alignment) && (alignment <= MaxAlignment));
 
+    /* Get hold of the current offset */
     secOffs = emitConsDsc.dsdOffs;
 
-    if (align)
+    if (alignment > 4)
     {
-        // Data can have any size but since alignment is deduced from the size there's no
-        // way to have a larger data size (e.g. 128) and request 4/8/16 byte alignment.
-        // As such, we restrict data above 16 bytes to be a multiple of 16 and assume 16-byte
-        // alignment. Alignment greater than 16 requires VM support (see ICorJitInfo::allocMem).
-        assert((size <= 16) || ((size % 16) == 0));
+        uint8_t zero[MaxAlignment] = {};
 
-        if (size >= 16)
-        {
-            emitConsDsc.align16 = true;
-        }
+        UNATIVE_OFFSET zeroSize  = alignment - (secOffs % alignment);
+        UNATIVE_OFFSET zeroAlign = 4;
 
-        while ((secOffs % size) != 0)
-        {
-            /* Need to skip 4 bytes to honor alignment */
-            /* Must allocate a dummy 4 byte integer */
-            int zero = 0;
-            emitDataGenBeg(4, false);
-            emitDataGenData(0, &zero, 4);
-            emitDataGenEnd();
-
-            /* Get the new secOffs */
-            secOffs = emitConsDsc.dsdOffs;
-        }
+        emitAnyConst(&zero, zeroSize, zeroAlign);
+        secOffs = emitConsDsc.dsdOffs;
     }
 
-    /* Advance the current offset */
+    assert((secOffs % alignment) == 0);
+    emitConsDsc.alignment = max(emitConsDsc.alignment, alignment);
```

Comment:
We track the maximum alignment requested here so we know what `CORJIT_ALLOCMEM_FLG_RODATA_*_FLAG` to use above.